src/arm-codegen.c

```diff
@@ -7,157 +7,184 @@ void emit(int code)
-            elf_code_idx += 28;
+        case OP_global_store:
+            elf_offset += 4;
             break;
         default:
-            break;
+            printf("Unknown OP\n");
+            abort();
         }
     }
+    /* jump to main */
+    elf_offset += 4;
 
-    elf_data_start = elf_code_start + elf_code_idx;
-    global_stack_size = FUNCS[0].stack_size;
-    block_lv = 0;
-    elf_code_idx = 0;
+    fn_t *fn;
+    for (fn = FUNC_LIST.head; fn; fn = fn->next) {
+        ph2_ir_t *flatten_ir;
 
-    /* insert entry, exit point and syscall manually */
-    elf_add_symbol("__start", strlen("__start"), elf_code_idx);
-    emit(__movw(__AL, __r8, global_stack_size));
-    emit(__movt(__AL, __r8, global_stack_size));
+        /* reserve stack */
+        flatten_ir = add_ph2_ir(OP_define);
+        flatten_ir->src0 = fn->func->stack_size;
+
+        basic_block_t *bb;
+        for (bb = fn->bbs; bb; bb = bb->rpo_next) {
+            bb->elf_offset = elf_offset;
+
+            if (bb == fn->bbs)
+                /* save ra, sp */
+                elf_offset += 16;
+
+            ph2_ir_t *inst;
+            for (inst = bb->ph2_ir_list.head; inst; inst = inst->next) {
+                switch (inst->op) {
+                case OP_load_constant:
+                    flatten_ir = add_ph2_ir(OP_load_constant);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+
+                    if (flatten_ir->src0 < 0)
+                        elf_offset += 12;
+                    else if (flatten_ir->src0 > 255)
+                        elf_offset += 8;
+                    else
+                        elf_offset += 4;
+
+                    break;
+                case OP_assign:
+                    if (inst->dest != inst->src0) {
+                        flatten_ir = add_ph2_ir(OP_assign);
+                        memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+                        elf_offset += 4;
+                    }
+                    break;
+                case OP_address_of:
+                    flatten_ir = add_ph2_ir(OP_address_of);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+
+                    if (flatten_ir->src0 > 255)
+                        elf_offset += 12;
+                    else
+                        elf_offset += 4;
+
+                    break;
+                case OP_load:
+                case OP_global_load:
+                    flatten_ir = add_ph2_ir(OP_address_of);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+
+                    if (flatten_ir->src0 > 255)
+                        elf_offset += 16;
+                    else
+                        elf_offset += 4;
+
+                    break;
+                case OP_store:
+                case OP_global_store:
+                    flatten_ir = add_ph2_ir(OP_address_of);
+                    memcpy(flatten_ir, inst, sizeof(ph2_ir_t));
+
+                    if (flatten_ir->src1 > 255)
```

Comment:
Add comments to explain the load/store model.

Reply:
I misunderstood the instruction in ARMv7-A manual. The immediate value uses 12 digits instead of 8 digits. I'll fix it.